### ISSUE TITLE
node-red-node-daemon: output object handling

### ISSUE THREAD

**Issue by ctr49** [node-red/node-red-nodes#263]
I have a daemon that produces json data. This can be handy as functions could access objects in the nested json like "msg.payload.object". But unfortunately, regardless of the input handling method (string, lines) my input json object is treated as string and quoted as result.

Am I missing something or is this the intended behaviour?

I created a workaround, not sure if that is acceptable though.

**Comment by dceejay** [comment]
does the json node not convert it back ok ?

**Comment by knolleary** [comment]
see also my comment on the associated pr #264

**Comment by ctr49** [comment]
I haven't tried it. Per my understanding the json node converts between json and object (and vice-versa). The current behaviour is to put a json as string value (with escaped quotation marks) rather than a nested json:
input:
{ "value": "001" }
current behavior (valid json, but value cannot be accessed directly using msg.payload.value):
{ "payload": "{ \"value\": \"001\" }" }
expected behavior (and actual behavior with my patch):
{ "payload": { "value": "001" } }

**Comment by knolleary** [comment]
Try it. It works :)

**Comment by ctr49** [comment]
But why should "daemon" mangle the input in the first place to have a another node clean things up?

**Comment by knolleary** [comment]
It doesn't mangle it. It puts the output received from the process into the 'payload' property of the message it sends. That is entirely consistent with how, by convention, every node operates.

**Comment by dceejay** [comment]
what gets received inside the daemon node is a stream of bytes from the stdout pipe of the child process. There is no concept of an object at that point. As Nick pointed out in  #264 , that could be an option - but why force every node to have that option when the JSON node already does what is needed.

**Comment by ctr49** [comment]
I agree that the approach to change the default behaviour might not be the best. I was looking at it from the perspective of "best effort" handling. If the output of the daemon is already a (valid) object, then why bother and modify it. But this would change the behaviour for existing users, so agreed, not a good idea.

However I still don't understand the concept why perfectly valid output (json object or javascript object) needs to be converted to a string (escaping the structure out of the object) just to use another node to fix it again.

**Comment by knolleary** [comment]
@ctr49: see the comment from @dceejay 

What you describe as 'perfectly valid output' is just a stream of raw bytes coming from the external process. That is what we are given to pass on. If we did nothing, you'd just get a raw buffer object containing those bytes. As it is, we provide the convenience to convert those raw bytes into a string.

The node is not converting a valid Object into a String. It is converting raw bytes to a string, which you may then be able to parse further into a JavaScript Object.

**Comment by ctr49** [comment]
That's just semantics, what I meant is "a perfectly valid string representation of an object". Instead of creating a nested object it just uses the string. If only raw output was desired there wouldn't be different options for input.
It's pointless to discuss this further, my takeaway
1. The current behaviour is intentional.
2. My patch was valid under the assumption that it is not intentional, so it changes default behaviour which is not good.
3. I still think it would be a good idea to have the **option** to directly emit a nested object (using a different approach than my patch though).

**Comment by knolleary** [comment]
Ok, I think we're on the same page. If the process passes us a perfectly valid string representation of an object then we do pass it on as-is - no mangling involved.

Current option is to use a JSON node to further parse that string into an actual object.

We could consider an explicit option on the node to do it for you, but the JSON node exists for this purpose and if we did it on this node, we'd set ourselves up for doing it everywhere.

Phew. Think we got there.
